### PR TITLE
Better logging for UHD components.

### PR DESCRIPTION
**Pull request by marcusmueller**
depends on #3339

Remove 'don't merge' as soon as that's merged.

### CODE REVIEW

File: gr-uhd/lib/usrp_sink_impl.cc [hunk 1]
```diff
@@ -691,6 +705,28 @@ void usrp_sink_impl::async_event_loop()
             pmt::pmt_t msg = pmt::cons(ASYNC_MSG_KEY, value);
             message_port_pub(ASYNC_MSGS_PORT_KEY, msg);
         }
+        if (sequence_error_occurred) {
+            GR_LOG_ERROR(d_logger,
```
[the remark this answers is not in the document]
Comment (marcusmueller) [reply gr-uhd/lib/usrp_sink_impl.cc:711]: Can do, but honestly, once you see a sequence error in UHD "classic", it's game over 90% of the time (not that UHD won't recover, but it's usually that you have broken networking equipment or a USB controller that resets randomly), so the "spamming" kind of reflects the severity.

File: gr-uhd/lib/usrp_sink_impl.cc [hunk 2]
```diff
@@ -691,6 +705,28 @@ void usrp_sink_impl::async_event_loop()
...
+                         "A sequence error occurred. This is typically indicative of "
+                         "network problems, or gross computational overload.")
```
[the remark this answers is not in the document]
Comment (marcusmueller) [reply gr-uhd/lib/usrp_sink_impl.cc:713]: yep, was looking for better words. I was extremely incompetent at finding one at that point ;)

File: gr-uhd/lib/usrp_sink_impl.cc [hunk 3]
```diff
@@ -691,6 +705,28 @@ void usrp_sink_impl::async_event_loop()
...
+        }
+        if (underflow_counter || time_error_counter) {
+            auto now = clock::now();
+            auto delta = now - last_emission;
+            last_emission = now;
+            if (delta > underflow_log_interval) {
+                auto ms =
+                    std::chrono::duration_cast<std::chrono::milliseconds>(delta).count();
+                if (underflow_counter) {
+                    GR_LOG_ERROR(d_logger, uflow_msg % ms % underflow_counter);
```
[the remark this answers is not in the document]
Comment (marcusmueller) [reply gr-uhd/lib/usrp_sink_impl.cc:723]: ahhh true, a logic bug on my side.

File: gr-uhd/lib/usrp_sink_impl.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbr0wn) [reply gr-uhd/lib/usrp_sink_impl.cc:711]: The scenario I was thinking of is when using SFP cables with NICs that don't work well together. It'll just keep dropping packets at the hardware, and you get spammed with S or D.

**Comment by marcusmueller** [comment]
A bit unusual to re-request review before CI is finished, but doing it, anyways, because: bedtime.

File: gr-uhd/lib/usrp_sink_impl.cc [hunk 4]
```diff
@@ -730,18 +749,22 @@ void usrp_sink_impl::async_event_loop()
         }
         if (metadata.event_code & md_t::EVENT_CODE_UNDERFLOW) {
             event_list = pmt::list_add(event_list, UNDERFLOW_KEY);
+            ++underflow_counter;
         }
         if (metadata.event_code & md_t::EVENT_CODE_UNDERFLOW_IN_PACKET) {
             event_list = pmt::list_add(event_list, UNDERFLOW_IN_PACKET_KEY);
```
Comment (mbr0wn) [inline gr-uhd/lib/usrp_sink_impl.cc:755]: This clause should also increase underflow_counter, right?

File: gr-uhd/lib/usrp_source_impl.h [hunk 5]
```diff
@@ -11,6 +11,8 @@
 #include "usrp_block_impl.h"
 #include <gnuradio/uhd/usrp_source.h>
 #include <uhd/convert.hpp>
+#include <boost/thread/mutex.hpp>
```
Comment (mbr0wn) [inline gr-uhd/lib/usrp_source_impl.h:14]: Is there a boost mutex in this file?

**Review by mbr0wn — COMMENTED** [review]
Looking good -- but please check out my last two comments/nits.

File: gr-uhd/lib/usrp_source_impl.h (hunk 5 above)
Comment (marcusmueller) [reply gr-uhd/lib/usrp_source_impl.h:14]: Oh, that must have been a mistake while rebasing.

Comment (marcusmueller) [reply gr-uhd/lib/usrp_source_impl.h:14]: Also fixing the other ICWYU omissions.

File: gr-uhd/lib/usrp_sink_impl.cc (hunk 4 above)
Comment (marcusmueller) [reply gr-uhd/lib/usrp_sink_impl.cc:755]: Indeed!

**Review by mbr0wn — APPROVED** [review]

**Comment by willcode** [comment]
@marcusmueller Are you still thinking this should go in 3.8? Backport to 3.9 drew in a few other PRs. I could use some help with the 3.8 port.

**Pull request merged**